hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java

```diff
@@ -412,19 +416,52 @@ private void onMigrate(
         final var nodeAddress = platform.getAddressBook().getAddress(selfId);
         final var selfNodeInfo = SelfNodeInfoImpl.of(nodeAddress, version);
         final var networkInfo = new NetworkInfoImpl(selfNodeInfo, platform, bootstrapConfigProvider);
-        for (final var registration : servicesRegistry.registrations()) {
+
+        logger.info("Migrating Entity ID Service as pre-requisite for other services");
+        final var entityIdRegistration = servicesRegistry.registrations().stream()
+                .filter(service -> EntityIdService.NAME.equals(service.service().getServiceName()))
+                .findFirst()
+                .orElseThrow();
+        final var entityIdRegistry = (MerkleSchemaRegistry) entityIdRegistration.registry();
+        entityIdRegistry.migrate(
+                state,
+                previousVersion,
+                currentVersion,
+                configProvider.getConfiguration(),
+                networkInfo,
+                backendThrottle,
+                // We call with null here because we're migrating the entity ID service itself
+                null);
+        // Now that the Entity ID Service is migrated, migrate the remaining services
+        for (final var registration : servicesRegistry.registrations().stream()
+                .filter(r -> !Objects.equals(entityIdRegistration, r))
+                .collect(Collectors.toSet())) {
             // FUTURE We should have metrics here to keep track of how long it takes to migrate each service
             final var service = registration.service();
             final var serviceName = service.getServiceName();
             logger.info("Migrating Service {}", serviceName);
             final var registry = (MerkleSchemaRegistry) registration.registry();
+
+            // The token service has a dependency on the entity ID service during genesis migrations, so we CAREFULLY
+            // create a different WritableStates specific to the entity ID service. The different WritableStates
+            // instances won't be able to see the changes made by each other, but there shouldn't be any conflicting
+            // changes. We'll inject this into the MigrationContext below to enable generation of entity IDs.
+            final var entityIdWritableStates = state.createWritableStates(EntityIdService.NAME);
+            final var entityIdStore = new WritableEntityIdStore(entityIdWritableStates);
+
             registry.migrate(
                     state,
                     previousVersion,
                     currentVersion,
                     configProvider.getConfiguration(),
                     networkInfo,
-                    backendThrottle);
+                    backendThrottle,
+                    requireNonNull(entityIdStore));
```

Comment:
I wonder if it is sufficient to pass `state` into `migrate()` and create the store there. It feels awkward to require `state` and a specific store here. The `registry` can create the `entityIdStore` only when needed.

I believe when you call a contructor, the result can never be `null`. Either it returns an instance or throws.

Reply:
This is definitely possible, but it makes MerkleRegistrySchemaTest really ugly. The functionality inside of the `Hedera` class – where the EntityIdService is migrated first, followed by all the other services – has to be applied to  ~20 tests, which otherwise throw an exception:
```
java.lang.IllegalArgumentException: There are no singleton states
```
What would you prefer? Keeping the entity ID store where it is, or going the test route?

Reply:
Keeping it as is for now..